### PR TITLE
First implementation of ITS3 geometry

### PR DESCRIPTION
**Pull request by mconcas**
Tagging @shahor02 @sawenzel @qgp for knowledge.
This is a draft PR to better see what would be the addition to the code.

Simulation works with ITS and ITS3, I just would like to be sure not to break anything outside the `Upgrade` directory.

### CODE REVIEW

File: DataFormats/Detectors/Common/include/DetectorsCommonDataFormats/DetID.h [hunk 1]
```diff
@@ -109,13 +110,14 @@ class DetID
 
   static constexpr const char* sDetNames[nDetectors + 1] = ///< defined detector names
 
-    {"ITS", "TPC", "TRD", "TOF", "PHS", "CPV", "EMC", "HMP", "MFT", "MCH", "MID", "ZDC", "FT0", "FV0", "FDD", "ACO", nullptr};
+    {"ITS", "TPC", "TRD", "TOF", "PHS", "CPV", "EMC", "HMP", "MFT", "MCH", "MID", "ZDC", "FT0", "FV0", "FDD", "ACO", "ITS3", nullptr};
```
Comment (sawenzel) [inline DataFormats/Detectors/Common/include/DetectorsCommonDataFormats/DetID.h:113]: I am wondering if we will run into problems with DPL headers. I remember that detector origins origins can only be 3 letter, or maybe I am wrong?

File: Detectors/Upgrade/ITS3/base/include/ITS3Base/ContainerFactory.h [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+class FairParSet;
+
+class FairContainer;
+
+namespace o2
+{
+namespace its3
+{
+class ContainerFactory : public FairContFact
```
Comment (sawenzel) [inline Detectors/Upgrade/ITS3/base/include/ITS3Base/ContainerFactory.h:28]: Do we still need this? We are not really using this container/parameter feature of FairRoot.
(So essentially I am wondering if the code compiles and runs without this).

File: Detectors/Upgrade/ITS3/base/include/ITS3Base/MisalignmentParameter.h [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+  TArrayD getRotZ() { return mRotZ; }
+  Int_t getNumberOfDetectors() { return mNumberOfDetectors; }
+
+ private:
+  TArrayD mShiftX;          ///< Array to hold the misalignment in x-direction
+  TArrayD mShiftY;          ///< Array to hold the misalignment in y-direction
+  TArrayD mShiftZ;          ///< Array to hold the misalignment in z-direction
+  TArrayD mRotX;            ///< Array to hold the rotation in x-direction
+  TArrayD mRotY;            ///< Array to hold the rotation in y-direction
+  TArrayD mRotZ;            ///< Array to hold the rotation in z-direction
```
Comment (sawenzel) [inline Detectors/Upgrade/ITS3/base/include/ITS3Base/MisalignmentParameter.h:58]: These types of ROOT containers are outdated. Consider using standard C++ containers.

File: Detectors/Upgrade/ITS3/macros/test/run_sim_its3.C [hunk 4]
```diff
@@ -0,0 +1,194 @@
+#if !defined(__CLING__) || defined(__ROOTCLING__)
+#include <Rtypes.h>
+#include <TSystem.h>
+#include <TMath.h>
```
Comment (sawenzel) [inline Detectors/Upgrade/ITS3/macros/test/run_sim_its3.C:4]: is this file really needed? If possible, consider directly using o2-sim.

File: Detectors/Upgrade/ITS3/simulation/include/ITS3Simulation/Detector.h [hunk 5]
```diff
@@ -0,0 +1,382 @@
...
+#ifndef ALICEO2_ITS3_DETECTOR_H_
+#define ALICEO2_ITS3_DETECTOR_H_
+
+#include <vector>                             // for vector
+#include "DetectorsBase/GeometryManager.h"    // for getSensID
+#include "DetectorsBase/Detector.h"           // for Detector
+#include "DetectorsCommonDataFormats/DetID.h" // for Detector
+#include "ITSMFTSimulation/Hit.h"             // for Hit
+#include "Rtypes.h"                           // for Int_t, Double_t, Float_t, Bool_t, etc
+#include "TArrayD.h"                          // for TArrayD
```
Comment (sawenzel) [inline Detectors/Upgrade/ITS3/simulation/include/ITS3Simulation/Detector.h:23]: is this needed?

File: Steer/src/O2MCApplication.cxx [hunk 6]
```diff
@@ -86,9 +86,7 @@ void O2MCApplicationBase::ConstructGeometry()
       mModIdToName[mod->GetModId()] = mod->GetName();
     }
   }
-
   FairMCApplication::ConstructGeometry();
-
   std::ofstream voltomodulefile("MCStepLoggerVolMap.dat");
```
Comment (sawenzel) [inline Steer/src/O2MCApplication.cxx:90]: why is this file touched? Should probably not be part of this PR.

File: macro/build_geometry.C [hunk 7]
```diff
@@ -24,6 +24,7 @@
 #include <Field/MagneticField.h>
 #include <TPCSimulation/Detector.h>
 #include <ITSSimulation/Detector.h>
+#include <ITS3Simulation/Detector.h>
```
Comment (sawenzel) [inline macro/build_geometry.C:27]: this will need a protection in future (#ifdef ENABLE_UPGRADE)

File: DataFormats/Detectors/Common/include/DetectorsCommonDataFormats/DetID.h (hunk 1 above)
Comment (shahor02) [reply DataFormats/Detectors/Common/include/DetectorsCommonDataFormats/DetID.h:113]: @sawenzel true, better have it as 3 chars, e.g. IT3.
Strictly speaking, the mapping from the DetID to getectorOrigin is done in the yet unmerged https://github.com/AliceO2Group/AliceO2/pull/3685: https://github.com/AliceO2Group/AliceO2/blob/c242bda0dc1367017b1491bb5f3a21cf14eb1575/DataFormats/Detectors/Common/include/DetectorsCommonDataFormats/DetID.h#L138-L141, and since we will never have 
ITS and ITS3 at the same time, we could use the same gDataOriginITS for both, but in many places in the code we use the name instead of the origin, so better to have them the same.

File: macro/build_geometry.C [hunk 8]
```diff
@@ -141,7 +142,7 @@ void build_geometry(FairRunSim* run = nullptr)
 
   // beam pipe
   if (isActivated("PIPE")) {
-    run->AddModule(new o2::passive::Pipe("PIPE", "Beam pipe"));
+    run->AddModule(new o2::passive::Pipe("PIPE", "Beam pipe", 1.6f, 0.05));
```
Comment (sawenzel) [inline macro/build_geometry.C:145]: what are these new parameters? If specific to upgrade --> needs a protection.

File: run/O2HitMerger.h [hunk 9]
```diff
@@ -664,7 +665,7 @@ void O2HitMerger::initDetInstances()
     auto active = std::find(modulelist.begin(), modulelist.end(), s) != modulelist.end();
     return active; };
 
-  mDetectorInstances.resize(DetID::Last);
+  mDetectorInstances.resize(DetID::nDetectors);
```
Comment (sawenzel) [inline run/O2HitMerger.h:668]: why this change?

File: DataFormats/Detectors/Common/include/DetectorsCommonDataFormats/DetID.h (hunk 1 above)
Comment (mconcas) [reply DataFormats/Detectors/Common/include/DetectorsCommonDataFormats/DetID.h:113]: > I am wondering if we will run into problems with DPL headers. I remember that detector origins origins can only be 3 letter, or maybe I am wrong?

Yes, I was also wondering, I'll make it `IT3`.

File: Detectors/Upgrade/ITS3/base/include/ITS3Base/ContainerFactory.h (hunk 2 above)
Comment (mconcas) [reply Detectors/Upgrade/ITS3/base/include/ITS3Base/ContainerFactory.h:28]: You mean the whole file, right?
Now that I notice the `ALICEO2_ITS_CONTAINERFACTORY_H_` symbol is the same created in ITS, I forgot to convert it. I guess I can try without it.

File: Detectors/Upgrade/ITS3/base/include/ITS3Base/MisalignmentParameter.h (hunk 3 above)
Comment (mconcas) [reply Detectors/Upgrade/ITS3/base/include/ITS3Base/MisalignmentParameter.h:58]: True, this should apply for the ITS too.
Perhaps I can do a later iteration changing both of them?

File: Detectors/Upgrade/ITS3/macros/test/run_sim_its3.C (hunk 4 above)
Comment (mconcas) [reply Detectors/Upgrade/ITS3/macros/test/run_sim_its3.C:4]: No, it is a leftover, I'll move it away.

File: Detectors/Upgrade/ITS3/simulation/include/ITS3Simulation/Detector.h (hunk 5 above)
Comment (mconcas) [reply Detectors/Upgrade/ITS3/simulation/include/ITS3Simulation/Detector.h:23]: `TArrayD.h`? not at all, here can be really removed. See also previous comment.

File: Steer/src/O2MCApplication.cxx (hunk 6 above)
Comment (mconcas) [reply Steer/src/O2MCApplication.cxx:90]: True, I'll fix it.

File: macro/build_geometry.C (hunk 8 above)
Comment (mconcas) [reply macro/build_geometry.C:145]: Indeed.

File: run/O2HitMerger.h (hunk 9 above)
Comment (mconcas) [reply run/O2HitMerger.h:668]: This I am not 100% sure, but when we add the ITS3 (with DetID=16) the total number of detectors becomes 17. 
The next loop index `i` goes from `DetID::First` (0) to `DetID::Last` (16) included and accesses the `mDetectorInstances[i]`.
If size of `mDetectorInstances` were `DetID::Last` (16) it means that `mDetectorInstances[16]` would not be something legal.
But maybe I missed something but if I remember correctly it did not work until I changed this.

File: run/O2HitMerger.h [hunk 10]
```diff
@@ -733,12 +734,16 @@ void O2HitMerger::initDetInstances()
...
+      mDetectorInstances[i] = std::move(std::make_unique<o2::its3::Detector>(true));
+      counter++;
+    }
 
     // init the detector specific output files
     initHitTreeAndOutFile(i);
   }
-  if (counter != DetID::Last) {
-    LOG(WARNING) << " O2HitMerger: Some Detectors are potentially missing in this initialization ";
+  if (counter != DetID::nDetectors) {
```
Comment (mconcas) [inline run/O2HitMerger.h:745]: I guess same applies here?

File: run/O2HitMerger.h [hunk 11]
```diff
@@ -733,12 +734,16 @@ void O2HitMerger::initDetInstances()
...
+    LOG(WARNING) << " O2HitMerger: Some Detectors are potentially missing in this initialization " << counter;
```
Comment (mconcas) [inline run/O2HitMerger.h:746]: And this is to be removed

**Comment by mconcas** [comment]
I'll wait for #3702 to be merged, in order to implement the `ENABLE_UPGRADES` guards.

File: DataFormats/Detectors/Common/include/DetectorsCommonDataFormats/DetID.h (hunk 1 above)
Comment (qgp) [reply DataFormats/Detectors/Common/include/DetectorsCommonDataFormats/DetID.h:113]: @shahor02 @mconcas Just to make sure: We will then consider ITS3 to comprise both the new IB and the old OB, correct? So, ITS3 will have to duplicate the OB-related geometry and code?

Comment (mconcas) [reply DataFormats/Detectors/Common/include/DetectorsCommonDataFormats/DetID.h:113]: Yes, it can be a struggle in case of modifications on the OB of the ITS, but at the same time minimises the probability to break something in production code.

**Comment by qgp** [comment]
> I'll wait for #3702 to be merged, in order to implement the `ENABLE_UPGRADES` guards.

It should be sufficient to add the `add_subdirectory` in `Detectors/Upgrades/CMakeLists.txt` which is already protected. Or do we need to protect anything outside of that part?

**Comment by mconcas** [comment]
> > I'll wait for #3702 to be merged, in order to implement the `ENABLE_UPGRADES` guards.
> 
> It should be sufficient to add the `add_subdirectory` in `Detectors/Upgrades/CMakeLists.txt` which is already protected. Or do we need to protect anything outside of that part?

Some other files, like the ones involved in extending the simulation tool have to be protected. Not a big deal at the moment.

File: DataFormats/Detectors/Common/include/DetectorsCommonDataFormats/DetID.h (hunk 1 above)
Comment (shahor02) [reply DataFormats/Detectors/Common/include/DetectorsCommonDataFormats/DetID.h:113]: @qgp yes, I would completely duplicate all layers anyway.

**Comment by qgp** [comment]
@mconcas I think you can go ahead and implement the updates using `ENABLE_UPGRADES`. Let me know if anything is missing.

**Comment by mconcas** [comment]
> @mconcas I think you can go ahead and implement the updates using `ENABLE_UPGRADES`. Let me know if anything is missing.

The `ENABLE_UPGRADES` option is a CMake Option, I am currently figuring out how to pass the definition of the preprocessor macro in a way that is accepted by the build, because at the moment it behave like `ENABLE_UPGRADES` macro is not passed.

**Comment by sawenzel** [comment]
Taking a look now. I think the cmake is just offering the option ENABLE_UPGRADE but not yet setting the compiler definitions.

**Comment by sawenzel** [comment]
@mconcas @qgp : This PR needs https://github.com/AliceO2Group/AliceO2/pull/3764 as well as https://github.com/alisw/alidist/pull/2296. Thereafter you should be able to compile this with
`ENABLE_UPGRADES=ON aliBuild build O2 --defaults o2`.

I would suggest to wait until these things are merged and then we can adjust this PR.

File: DataFormats/Headers/include/Headers/DataHeader.h [hunk 12]
```diff
@@ -42,6 +42,8 @@
 #include "MemoryResources/Types.h"
 #include <cerrno>
 
+// #define ENABLE_UPGRADES
```
Comment (sawenzel) [inline DataFormats/Headers/include/Headers/DataHeader.h:45]: this can be removed

File: Detectors/Base/src/GeometryManager.cxx [hunk 13]
```diff
@@ -51,7 +51,6 @@ Bool_t GeometryManager::getOriginalMatrix(const char* symname, TGeoHMatrix& m)
 
   if (!gGeoManager || !gGeoManager->IsClosed()) {
     LOG(ERROR) << "No active geometry or geometry not yet closed!";
-    ;
```
Comment (sawenzel) [inline Detectors/Base/src/GeometryManager.cxx:54]: should not be part of this PR

File: Steer/src/O2MCApplication.cxx [hunk 14]
```diff
@@ -86,7 +86,6 @@ void O2MCApplicationBase::ConstructGeometry()
       mModIdToName[mod->GetModId()] = mod->GetName();
     }
   }
-
```
Comment (sawenzel) [inline Steer/src/O2MCApplication.cxx:89]: should not be part of this PR

File: cmake/O2DefineOptions.cmake [hunk 15]
```diff
@@ -28,5 +28,5 @@ function(o2_define_options)
   # for the complete picture of how BUILD_SIMULATION is handled see
   # ../dependencies/O2SimulationDependencies.cmake
 
-  option(ENABLE_UPGRADES "Enable detectors for upgrades" OFF)
```
Comment (sawenzel) [inline cmake/O2DefineOptions.cmake:31]: should not be changed. You'll be able to set this from the outside

File: macro/CMakeLists.txt [hunk 16]
```diff
@@ -117,24 +117,46 @@ o2_add_test_root_macro(analyzeOriginHits.C
...
+                                               O2::TOFSimulation
+                                               O2::TRDSimulation
+                                               O2::FT0Simulation
+                                               O2::FV0Simulation
+                                               O2::FDDSimulation
+                                               O2::HMPIDSimulation
+                                               O2::PHOSSimulation
+                                               O2::CPVSimulation
+                                               O2::ZDCSimulation)
+endif()
```
Comment (sawenzel) [inline macro/CMakeLists.txt:159]: this is a bit long. Maybe there is a way to construct a list and pass the list to the function. @aphecetche ?

File: run/CMakeLists.txt [hunk 17]
```diff
@@ -36,6 +38,32 @@ target_link_libraries(allsim
...
+                                O2::MIDSimulation
+                                O2::PHOSSimulation
+                                O2::FT0Simulation
+                                O2::TOFSimulation
+                                O2::TPCSimulation
+                                O2::TRDSimulation
+                                O2::FV0Simulation
+                                O2::ZDCSimulation
+                                O2::Generators)
+endif()
```
Comment (sawenzel) [inline run/CMakeLists.txt:66]: same comment as above

File: run/O2HitMerger.h (hunk 9 above)
Comment (sawenzel) [reply run/O2HitMerger.h:671]: I guess you need to adjust DetID::Last when you add some detector. The logic should not be changed here.

Comment (sawenzel) [reply run/O2HitMerger.h:671]: Ok. Here the change seems ok. I just checked.

**Comment by qgp** [comment]
@mconcas For the targets that need the definition, you should add something like
```
if (ENABLE_UPGRADES)
  target_compile_definitions(<target> ENABLE_UPGRADES)
endif()
```
Or did I miss what you were asking for?

**Comment by mconcas** [comment]
> @mconcas For the targets that need the definition, you should add something like
> 
> ```
> if (ENABLE_UPGRADES)
>   target_compile_definitions(<target> PUBLIC ENABLE_UPGRADES)
> endif()
> ```
> 
> Or did I miss what you were asking for?



> @mconcas For the targets that need the definition, you should add something like
> 
> ```
> if (ENABLE_UPGRADES)
>   target_compile_definitions(<target> PUBLIC ENABLE_UPGRADES)
> endif()
> ```
> 
> Or did I miss what you were asking for?

That is also something I was looking for (i tried now to put it [here](https://github.com/AliceO2Group/AliceO2/pull/3699/files#diff-749af43633a7c991870fc6212e2114beR16)).
The actual point is that the option(...) does not seem to play how I expect, I am doing some test now,  the build directory everytime. No success at the moment.

**Comment by sawenzel** [comment]
> @mconcas For the targets that need the definition, you should add something like
> 
> ```
> if (ENABLE_UPGRADES)
>   target_compile_definitions(<target> PUBLIC ENABLE_UPGRADES)
> endif()
> ```
> 
> Or did I miss what you were asking for?

Please note that this is no longer necessary. Upgrade macro definitions are now enabled in the main CMake file.

File: cmake/O2DefineOptions.cmake (hunk 15 above)
Comment (mconcas) [reply cmake/O2DefineOptions.cmake:31]: Option has been removed

Comment (sawenzel) [reply cmake/O2DefineOptions.cmake:31]: sorry. There was a misunderstanding. The declaration of this option is still necessary here (in my understanding). Otherwise you can't modify from the outside. What I wanted to say is maybe that this file should not be touched since the option is available correctly on the dev branch already.

Comment (mconcas) [reply cmake/O2DefineOptions.cmake]: ok, Reverted

**Comment by qgp** [comment]
@sawenzel @mconcas Ok, so we can globally assume to use `#ifdef ENABLE_UPGRADES` as guard for upgrade only stuff.

**Comment by sawenzel** [comment]
I guess it can and should be documented in a dedicated `Run4WhateverUpgrade.md` file.

**Comment by qgp** [comment]
> I guess it can and should be documented in a dedicated `Run4WhateverUpgrade.md` file.

Sounds good, I will take care of adding such a file in `doc/`.

**Comment by sawenzel** [comment]
So shall we go ahead merging this PR?

**Comment by mconcas** [comment]
Fine to me
